src/SB/Game/zNPCGlyph.cpp

```diff
@@ -23,14 +122,492 @@ void zNPCGlyph_ScenePostInit()
+    {
+        mdl_glyph = xModelInstanceAlloc(model_data, NULL, 0, 0, NULL);
+        if (mdl_glyph != NULL)
+        {
+            xMat4x3 frame;
+            xMat3x3Euler(&frame, &vec);
+            xVec3Copy(&frame.pos, &g_O3);
+            frame.flags = 0;
+            xModelSetFrame(mdl_glyph, &frame);
+        }
+    }
+
+    Reset();
+}
+
+void NPCGlyph::Kill()
+{
+    if (mdl_glyph != NULL)
+    {
+        xModelInstanceFree(mdl_glyph);
+    }
+    typ_glyph = NPC_GLYPH_UNKNOWN;
     flg_glyph = 0;
-    pos_glyph = g_O3;
-    vel_glyph = g_O3;
-    tmr_glyph = lbl_803CF948;
+    xVec3Copy(&pos_glyph, &g_O3);
+    xVec3Copy(&vel_glyph, &g_O3);
+    mdl_glyph = NULL;
+}
+
+void NPCGlyph::Render()
+{
+    xModelInstance* model = mdl_glyph;
+    if (model == NULL)
+    {
+        return;
+    }
+
+    if (flg_glyph & (1 << 6))
+    {
+        xVec3Copy(&model->Scale, &scl_glyph);
+        xModelRender(mdl_glyph);
+        xVec3Copy(&mdl_glyph->Scale, &g_Onez);
+    }
+    else
+    {
+        xModelRender(model);
+    }
+}
+
+void NPCGlyph::Enable(S32 ison)
+{
+    if (ison)
+    {
+        flg_glyph |= (1 << 1);
+    }
+    else
+    {
+        flg_glyph &= ~(1 << 1);
+    }
+}
+
+void NPCGlyph::Discard()
+{
+    en_npcglyph type = typ_glyph;
+    if (type > NPC_GLYPH_UNKNOWN && type < NPC_GLYPH_NOMORE)
+    {
+        g_cnt_activeGlyphs[type]--;
+        g_cnt_activeGlyphs[typ_glyph] = MAX(0, g_cnt_activeGlyphs[typ_glyph]);
+    }
+
+    flg_glyph = 0;
+}
+
+void NPCGlyph::PosSet(xVec3* pos)
+{
+    if (pos != NULL)
+    {
+        xVec3Copy(&pos_glyph, pos);
+    }
+    else
+    {
+        xVec3Copy(&pos_glyph, &g_O3);
+    }
+
+    if (mdl_glyph != NULL)
+    {
+        xVec3Copy((xVec3*)&mdl_glyph->Mat->pos, &pos_glyph);
+    }
+}
+
+void NPCGlyph::VelSet(xVec3* vel)
+{
+    if (vel != NULL)
+    {
+        xVec3Copy(&vel_glyph, vel);
+    }
+    else
+    {
+        xVec3Copy(&vel_glyph, &g_O3);
+    }
+
+    if (vel == NULL)
+    {
+        flg_glyph &= ~(1 << 2);
+    }
+    else if (xVec3Length2(vel) > 0.0f)
+    {
+        flg_glyph |= (1 << 2);
+    }
+    else
+    {
+        flg_glyph &= ~(1 << 2);
+    }
+}
+
+void NPCGlyph::ScaleSet(xVec3* scale)
+{
+    if (scale != NULL)
+    {
+        xVec3Copy(&scl_glyph, scale);
+        flg_glyph |= (1 << 6);
+    }
+    else
+    {
+        xVec3 one = { 1.0f, 1.0f, 1.0f };
+        xVec3Copy(&scl_glyph, &one);
+        flg_glyph &= ~(1 << 6);
+    }
+}
+
+void NPCGlyph::RotSet(xVec3* ang, S32 doautospin)
+{
+    xMat3x3 mat_rot = {};
+    xMat3x3Euler(&mat_rot, ang);
+    RotSet(&mat_rot, doautospin);
+}
+
+void NPCGlyph::RotSet(xMat3x3* mat_rot, S32 doautospin)
+{
+    xMat3x3Copy(&rot_glyph, mat_rot);
+    if (doautospin)
+    {
+        flg_glyph |= (1 << 5);
+    }
+    else
+    {
+        flg_glyph &= ~(1 << 5);
+    }
+}
+
+void NPCGlyph::VelSet_Shiny()
+{
+    tmr_glyph = -1.0f;
+    xVec3 local_18 = {};
+
+    {
+        xVec3 temp;
+        // Do we use sph here?
```

Comment:
I've been using cylinder for the player but I'm not sure if that's actually correct yet.